Code/Framework/AzCore/Tests/Math/IntersectionTests.cpp

```diff
@@ -496,109 +496,109 @@ namespace UnitTest
         EXPECT_GT(numMisses, 0);
     }
 
-#if AZ_TRAIT_DISABLE_FAILED_MATH_TESTS
-    TEST_F(MATH_IntersectSegmentTriangleTest, DISABLED_RegressionTestForSpecificSegmentsAndTriangles)
-#else
-    TEST_F(MATH_IntersectSegmentTriangleTest, RegressionTestForSpecificSegmentsAndTriangles)
-#endif
-    {
-        // This unit test is set up to validate regressions of erroneous segment/triangle intersection test results.
-        // All the failures that have been found so far have been false negatives or false positives that occur on triangle
-        // edges or vertices.
+     struct RayTriangleTest
+    {
+        AZ::Vector3 m_rayStart;
+        AZ::Vector3 m_rayEnd;
+        AZ::Vector3 m_triVerts[3];
+        bool m_shouldHit;
+        float m_hitDistance;
+        AZ::Vector3 m_hitNormal;
+    };
 
-        struct RayTriangleTest
+    static const RayTriangleTest RayTriangleTestParams[] =
+    {
+        // This failure came from a vegetation system raycast to a triangle in a sphere model.
         {
-            AZ::Vector3 m_rayStart;
-            AZ::Vector3 m_rayEnd;
-            AZ::Vector3 m_triVerts[3];
-            bool m_shouldHit;
-            float m_hitDistance;
-            AZ::Vector3 m_hitNormal;
-        };
+            { 0.0, 0.16, 1.01948643 }, { 0.0, 0.16, -0.0194873810 }, // segment
+            { { 0.0294727981, 0.18608, 0.9605 }, { 0.0, 0.18842, 0.9605 }, { 0.0, 0.15215, 0.9741 } }, // triangle
+            true, 0.0465169623, { 0.0278645027, 0.350958735, 0.935976326 }  // expected results
+        },
 
-        RayTriangleTest tests[] =
+        // These failures were generated from purely random segment/triangle combinations that were cross-checked against
+        // multiple algorithms in the "CompareSegmentTriangleIntersectionMethods" unit test.
         {
-            // This failure came from a vegetation system raycast to a triangle in a sphere model.
-            {
-                { 0.0, 0.16, 1.01948643 }, { 0.0, 0.16, -0.0194873810 }, // segment
-                { { 0.0294727981, 0.18608, 0.9605 }, { 0.0, 0.18842, 0.9605 }, { 0.0, 0.15215, 0.9741 } }, // triangle
-                true, 0.0465169623, { 0.0278645027, 0.350958735, 0.935976326 }  // expected results
-            },
+            { 962.503540, -788.401978, -320.390747 }, { -838.229004, 555.508301, 748.038086 }, // segment
+            { { -353.655212, 579.209229, -827.134216 }, { 358.185913, 346.845093, 729.328979 }, { 222.018555, -358.014160, -4.521179 } }, // triangle
+            true, 0.399751157, { 0.899101138, 0.220195562, -0.378326714 } // expected results
```

Comment:
aside: Packing expected results into a test fixture for a test can turn into a maintenance anti-pattern. Doesn't look like it's currently creating a problem, but be wary about the value of reuse versus any penalties to simplicity and readability.

Reply:
Since I'm the one who wrote these tests, I'll justify them. :) 

Our previous line/triangle intersection algorithms had precision issues where they could very occasionally fail with tiny triangles, rays that intersect edges or vertices, etc. These cases were extremely hard to catch, and I'm not sure how to create simple reproductions. To verify that the new algorithm didn't have these problems, I introduced a few sets of tests in here. One test fires a set of deterministic rays through the center of a highly-tesselated icosphere and verifies that there are always exactly 2 hits. One test runs all three line/triangle algorithms across a small set of deterministically-random rays and triangles and verifies that all three algorithms match in their results. Finally, this test takes a specific set of test cases that were discovered to fail with the previous algorithms and verifies them against the new algorithm to ensure that they succeed. This particular case was generated from running the three-algorithm-comparison test 10 million times and capturing some of the failing results that show up when you run that many combinations. (Other cases in this test were caught from failed Editor raycasts and failed vegetation system raycasts that occurred "in the wild".)

I would argue that this specific test has just proven its worth, because it did exactly what it's supposed to do and caught failures introduced into the system via the fast-math flag. By having a successful positive failure and no false failures yet, it feels pretty worthwhile to keep around and makes me super happy that I added it in the first place. :) 

I also 100% recognize that when it _does_ fail, there's no good way of debugging what's going on because of the mysteriousness of the numbers. If you can find a way to improve the understandability here without harming the reliability of the test, I'm all for it!

Reply:
Definitely agree that the caught failure has proved useful! As far as making the tests more linear, it would probably involve deconvoluting the different expectations into more-specific tests. Since they're all segment-to-triangle tests, it would probably need to be related to the edge case (results) it is trying to verify. Then each specific test could more quickly inform what failed.

Since I can't think of a better way to classify triangle data or edge cases, another thing that could help is to pack a string descriptor of each case along with the data. This can add context via `EXPECT_EQ(test.m_shouldHit, result) << test.m_caseDescriptor;` to ramp up whoever comes in to debug it. It would also document why each set of data contains what it does.

Reply:
I definitely like taking actual failure cases from actual production situations, and encoding them as tests. Especially for mathematical operations such as this. That said, it is a lot of magic numbers so lots of comments are encouraged.